# Tutorials for analysis of low-coverage whole genome sequencing data

These tutorials were developed by Matteo Fumagalli, Arne Jacobs, Nicolas Lou, and Nina Overgaard Therkildsen. An earlier version was used in the [Physalia course on Population genomic inference from low-coverage whole-genome sequencing data](https://www.physalia-courses.org/courses-workshops/course64/), Oct 19-22 2020

* [Tutorial 1: Data processing](https://github.com/nt246/lcwgs-guide-tutorial/blob/main/tutorial1_data_processing/markdowns/data_processing.md)

* [Tutorial 2: Genotype and SNP calling](https://github.com/nt246/lcwgs-guide-tutorial/blob/main/tutorial2_genotype_snp_calling/markdowns/00_introduction.md)

* [Tutorial 3a: Linkage disequilibrium](https://github.com/nt246/lcwgs-guide-tutorial/blob/main/tutorial3_ld_popstructure/markdowns/ld.md)

* [Tutorial 3b: Population structure](https://github.com/nt246/lcwgs-guide-tutorial/blob/main/tutorial3_ld_popstructure/markdowns/day3_PCA_Admixture_practicals.md)

* [Tutorial 4: Natural selection](https://github.com/nt246/lcwgs-guide-tutorial/blob/main/tutorial4_selection/markdowns/01_sfs.md)


<br>
<br>

**Acknowledgements**: This course material was developed with support from an NSF grant to NOT (OCE‚Äê1756316).

<br>
<br>

**Other resources**: 

Please check out our [beginner's guide to low-coverage whole gneome sequencing](https://doi.org/10.1111/mec.16077).

We also maintain two GitHub repos that host a collection of scripts for the efficient high-throughput [processing](https://github.com/therkildsen-lab/data-processing) and [analysis](https://github.com/therkildsen-lab/genomic-data-analysis) of lcWGS data.


